languages/rust/src/host/mod.rs

```diff
@@ -0,0 +1,476 @@
+}
+
+impl Host {
+    pub fn new(polar: Weak<crate::PolarCore>) -> Self {
+        let mut host = Self {
+            class_names: HashMap::new(),
+            classes: HashMap::new(),
+            instances: HashMap::new(),
+            polar,
+        };
+        let type_class = type_class();
+        let name = Name("Type".to_string());
+        host.class_names.insert(type_class.type_id, name.clone());
+        host.classes.insert(name, type_class);
+        host
+    }
+
+    pub fn type_class(&mut self) -> &mut Class {
+        self.classes.get_mut(&Name("Type".to_string())).unwrap()
+    }
+
+    pub fn get_class(&self, name: &Name) -> Option<&Class> {
+        self.classes.get(name)
+    }
+
+    pub fn get_class_from_type<C: 'static>(&self) -> Option<&Class> {
+        self.class_names
+            .get(&std::any::TypeId::of::<C>())
+            .and_then(|name| self.get_class(name))
+    }
+
+    pub fn get_class_mut(&mut self, name: &Name) -> Option<&mut Class> {
+        self.classes.get_mut(name)
+    }
+
+    /// Add the class to the host classes
+    ///
+    /// Returns an instance of `Type` for this class.
+    pub fn cache_class(&mut self, class: Class, name: Name) -> Term {
+        self.class_names.insert(class.type_id, name.clone());
+        self.classes.insert(name, class.clone());
+
+        let type_class = self.type_class();
+        for method_name in class.class_methods.keys() {
+            type_class
+                .instance_methods
+                .entry(method_name.clone())
+                .or_insert_with(|| {
+                    crate::host::InstanceMethod::from_class_method(method_name.clone())
+                });
+        }
+        let repr = format!("type<{}>", class.name);
+        let instance = type_class.cast_to_instance(class);
+        let instance = self.cache_instance(instance, None);
+        Term::new_from_ffi(Value::ExternalInstance(ExternalInstance {
+            constructor: None,
+            repr: Some(repr),
+            instance_id: instance,
+        }))
+    }
+
+    pub fn get_instance(&self, id: u64) -> Option<&Instance> {
+        self.instances.get(&id)
+    }
+
+    pub fn cache_instance(&mut self, instance: Instance, id: Option<u64>) -> u64 {
+        let id = id.unwrap_or_else(|| self.polar.upgrade().unwrap().get_external_id());
+        self.instances.insert(id, instance);
+        id
+    }
+
+    pub fn make_instance(&mut self, name: &Name, fields: Vec<Term>, id: u64) {
+        let class = self.get_class(name).unwrap().clone();
+        debug_assert!(self.instances.get(&id).is_none());
+        let fields = fields; // TODO: use
+        let instance = class.init(fields, self);
+        self.cache_instance(instance, Some(id));
+    }
+
+    pub fn unify(&self, left: u64, right: u64) -> bool {
+        let _left = self.get_instance(left).unwrap();
+        let _right = self.get_instance(right).unwrap();
+        todo!("left == right")
```

Comment:
Note: not implemented

Reply:
Do we have a plan for how to implement this? It doesn't seem like we actually are able to downcast into the real underlying type in order to call the `Eq` trait.

Reply:
We can follow the same approach as some of the similar methods, have an `eq_check: Arc<Fn(...)>` stored on the class which downcasts the args and checks equality. 
If we wanted to do this in a more generic way, we'll need to look into downcasting to trait objects using something like this: https://crates.io/crates/query_interface